File: tests/pytest/db_cleanup.py
"""Delete pytest-created users from Postgres so the DB stays clean."""

from __future__ import annotations

import os
from pathlib import Path

import psycopg


def load_database_url() -> str:
    url = os.getenv("DATABASE_URL")
    if url:
        return url.split("?")[0]

    env_path = Path(__file__).resolve().parents[2] / ".env"
    if env_path.exists():
        for line in env_path.read_text(encoding="utf-8").splitlines():
            stripped = line.strip()
            if not stripped or stripped.startswith("#") or "=" not in stripped:
                continue
            key, _, value = stripped.partition("=")
            if key.strip() == "DATABASE_URL":
                return value.strip().strip('"').strip("'").split("?")[0]

    raise RuntimeError(
        "DATABASE_URL not found. Set it in the environment or .env "
        "(same value the Next.js app uses)."
    )


def delete_users_by_emails(emails: list[str]) -> int:
    """Delete users (and cascaded rows) matching the given emails. Returns count deleted."""
    normalized = sorted({email.strip().lower() for email in emails if email and email.strip()})
    if not normalized:
        return 0

    with psycopg.connect(load_database_url()) as conn:
        with conn.cursor() as cur:
            cur.execute('DELETE FROM "User" WHERE lower(email) = ANY(%s)', (normalized,))
            deleted = cur.rowcount
        conn.commit()
    return deleted


def delete_events_by_ids(event_ids: list[str]) -> int:
    """Delete events (and cascaded Task/Contribution/CountMeIn rows) created during a test."""
    ids = sorted({e for e in event_ids if e})
    if not ids:
        return 0

    with psycopg.connect(load_database_url()) as conn:
        with conn.cursor() as cur:
            cur.execute('DELETE FROM "Event" WHERE id = ANY(%s)', (ids,))
            deleted = cur.rowcount
        conn.commit()
    return deleted


def delete_memberships_by_ids(membership_ids: list[str]) -> int:
    """Delete memberships created during a test (e.g. join requests)."""
    ids = sorted({m for m in membership_ids if m})
    if not ids:
        return 0

    with psycopg.connect(load_database_url()) as conn:
        with conn.cursor() as cur:
            cur.execute('DELETE FROM "Membership" WHERE id = ANY(%s)', (ids,))
            deleted = cur.rowcount
        conn.commit()
    return deleted

File: tests/integration/db-cleanup.ts
import { prisma } from "@/backend/db/prisma";

/** Deletes users (and cascaded rows) matching the given emails. Returns the count deleted. */
export async function deleteUsersByEmails(emails: string[]): Promise<number> {
  const normalized = [...new Set(emails.map((e) => e.trim().toLowerCase()).filter(Boolean))];
  if (normalized.length === 0) return 0;
  const result = await prisma.user.deleteMany({ where: { email: { in: normalized } } });
  return result.count;
}

/** Deletes events (and cascaded Task/Contribution/CountMeIn rows) created during a test. */
export async function deleteEventsByIds(eventIds: string[]): Promise<number> {
  const ids = [...new Set(eventIds.filter(Boolean))];
  if (ids.length === 0) return 0;
  const result = await prisma.event.deleteMany({ where: { id: { in: ids } } });
  return result.count;
}

/** Deletes memberships created during a test (e.g. join requests). */
export async function deleteMembershipsByIds(membershipIds: string[]): Promise<number> {
  const ids = [...new Set(membershipIds.filter(Boolean))];
  if (ids.length === 0) return 0;
  const result = await prisma.membership.deleteMany({ where: { id: { in: ids } } });
  return result.count;
}

/** Deletes announcements by id (e.g. Ask Sangam / domain integration fixtures). */
export async function deleteAnnouncementsByIds(announcementIds: string[]): Promise<number> {
  const ids = [...new Set(announcementIds.filter(Boolean))];
  if (ids.length === 0) return 0;
  const result = await prisma.announcement.deleteMany({ where: { id: { in: ids } } });
  return result.count;
}

/** Deletes tasks by id without touching their parent events. */
export async function deleteTasksByIds(taskIds: string[]): Promise<number> {
  const ids = [...new Set(taskIds.filter(Boolean))];
  if (ids.length === 0) return 0;
  const result = await prisma.task.deleteMany({ where: { id: { in: ids } } });
  return result.count;
}
